### PR TITLE
Orthographic sensor plugin

### PR DESCRIPTION
**Pull request by joeylitalien**
## Description

This PR adds an orthographic camera to the collection of sensors. This is a simple modification of the `PerspectiveCamera` where the projection was changed to orthographic (and rays are mapped accordingly). The documentation header follows [Mitsuba 0.6](https://github.com/mitsuba-renderer/mitsuba/blob/master/src/sensors/orthographic.cpp).

## Testing

A simple test script is provided, but may need more as it is just a distilled version of `test_perspective.py` without the FOV bits. Moreover, documentation build needs to be checked as I couldn't get Sphinx to work due to some version error with `sphinx-build` on my end.

### Example renders for docs
Below are example test scenes for the docs, adapted from `resources/data/docs/scenes/sensor_perspective.xml` and `resources/data/docs/scenes/integrator_depth_3.xml` respectively. Both are rendered at 1024 SPP and a camera scaling is provided for reproducibility.

Material test ball      |  Cornell box
:-------------------------:|:-------------------------:
  |  
`<scale x="2" y="2"/>` | `<scale x="300" y="300"/>`

## Checklist

- [x] My code follows the [style guidelines](https://mitsuba2.readthedocs.io/en/latest/src/developer_guide/intro.html#introduction) of this project
- [x] My changes generate no new warnings
- [x] My code also compiles for `gpu_*` and `packet_*` variants. If you can't test this, please leave below
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I cleaned the commit history and removed any "Merge" commits
- [x] I give permission that the Mitsuba 2 project may redistribute my contributions under the terms of its [license](https://github.com/mitsuba-renderer/mitsuba2/blob/master/LICENSE)

### CODE REVIEW

**Comment by Speierers** [comment]
Hi @joeylitalien ,

Awesome work!

Could you please open a PR in the `data` repository with the doc files.

Btw, it would also be great to add a rendering test (Add a scene in `resources/data/tests/scene/` and run `python3 ../src/librender/tests/test_render.py` to generate the reference images)

Then I will be able to merge the two PRs.

Thanks!

**Comment by joeylitalien** [comment]
Hi @Speierers,

See PR [#9](https://github.com/mitsuba-renderer/mitsuba-data/pull/8) of the `data` repository for the doc files.

Thanks!

**Comment by joeylitalien** [comment]
Ugh, for some reason I was on the `pathreparam-optix7` branch, which is not in sync with `master` for a lot of things including the sensors and the docs. Note that there is some discrepancy in the doc scenes (e.g. the Cornell box meshes are gone, even though they are used in some of the scenes). I will push the fix once this is handled and thoroughly tested. Apologies for this!

**Comment by Speierers** [comment]
Hi @joeylitalien ,

I have merged your PR on `resources`. Could you please update the ` resources/data` submodule (e.g. `git add resources/data && git commit`) and squash the commits of this PR into a single commit?

Then we can close this PR.

Thanks!

**Comment by standardisedfitting** [comment]
Hi, 

@Speierers thanks for pointing me to this PR from #274. I'm happy to get some validation on the plugin. 

@joeylitalien great that you also worked on an orthographic camera!
I noticed a couple of differences between our implementations.

In your ``update_camera_transforms()``-function are you setting the ``m_aspect`` variable and the ``m_needs_sample_3``?

Also I use ``transform_affine`` to compute the sample position in camera space while you multiply the transform. I think this might be the same thing in this case though.

Lastly, our ray differentials differ. Please correct me if I am mistaken, but I think the direction differentials should stay constant, while the origin diffrentials change based on the point.

I hope this is constructive.

**Comment by chenzhekl** [comment]
Hi all, I don‘t mean to rush anybody. I'm just wondering if there is still a chance to merge this PR, as it seems to have been on hold for a long time.

File: src/sensors/orthographic.cpp [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    ScalarTransform4f m_camera_to_sample;
+    ScalarTransform4f m_sample_to_camera;
+    ScalarBoundingBox2f m_image_rect;
+    ScalarFloat m_normalization;
+    ScalarVector3f m_dx, m_dy;
+};
+
+MTS_IMPLEMENT_CLASS_VARIANT(OrthographicCamera, ProjectiveCamera)
+MTS_EXPORT_PLUGIN(OrthographicCamera, "Orthographic Camera");
+NAMESPACE_END(mitsuba)
```
Comment (Speierers) [inline src/sensors/orthographic.cpp:200]: Would be great if you could add an empty line at the end of this file.

File: src/sensors/tests/test_orthographic.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    # Deltas in screen space
+    dx = 1.0 / camera.film().crop_size().x
+    dy = 1.0 / camera.film().crop_size().y
+
+    # Sample the rays by offsetting the position_sample with the deltas
+    ray_dx, _ = camera.sample_ray_differential(0, 0, [0.5 + dx, 0.5], 0)
+    ray_dy, _ = camera.sample_ray_differential(0, 0, [0.5, 0.5 + dy], 0)
+
+    assert ek.allclose(ray_dx.d, ray_center.d_x)
+    assert ek.allclose(ray_dy.d, ray_center.d_y)
```
Comment (Speierers) [inline src/sensors/tests/test_orthographic.py:121]: ditto

File: src/sensors/orthographic.cpp [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        ray.d = normalize(trafo * Vector3f(0, 0, 1));
+        ray.update();
+
+        ray.o_x = ray.o_y = ray.o;
+
+        ray.d_x = trafo * normalize(Vector3f(near_p) + m_dx);
+        ray.d_y = trafo * normalize(Vector3f(near_p) + m_dy);
+        ray.has_differentials = true;
+
+        return std::make_pair(ray, wav_weight);
```
Comment (Speierers) [inline src/sensors/orthographic.cpp:152]: Please use the following syntactic sugar 😉 
```
return { ray, wav_weight };
```

File: src/sensors/orthographic.cpp [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+        ray.mint = m_near_clip;
+        ray.maxt = m_far_clip;
+
+        auto trafo = m_world_transform->eval(ray.time, active);
+        ray.o = trafo.transform_affine(Point3f(near_p.x(), near_p.y(), 0.f));
+        ray.d = normalize(trafo * Vector3f(0, 0, 1));
+        ray.update();
+
+        return std::make_pair(ray, wav_weight);
```
Comment (Speierers) [inline src/sensors/orthographic.cpp:121]: Same here:
```
return { ray, wav_weight };
```

**Review by Speierers — COMMENTED** [review]
Hi @joeylitalien ,

Sorry for the delay, I haven't had the time to look into this PR for a while. 

Could you please address those few comments, then we should be able to merge this onto `master` and I will port those commits to the `next` branch 🚀

**Comment by hyperknot** [comment]
I've just finished upgrading a scene from 0.6 to 2.x only to have been surprised by the following error:
```
[xml.cpp:1047]   [PluginManager] Plugin "orthographic" not found!
```

Is there any chance that this would get merged in?

**Comment by Speierers** [comment]
This will be merged in the upcoming release.

**Comment by marcoromelli** [comment]
@Speierers It's been a while since your last comment. May I ask when is the upcoming release planned?

**Comment by Speierers** [comment]
@marcoromelli we are hoping to be able to release the next version of the codebase in the upcoming months.

**Comment by Speierers** [comment]
This PR was merged on the new codebase.

**Comment by marcoromelli** [comment]
@Speierers Probably a stupid question but, where is the "new codebase"? I tried to answer myself for the last hour but couldn't figure it out.

**Comment by njroussel** [comment]
@marcoromelli Hi, the new version @Speierers mentions has not been publicly released yet. It should be released to the public in the next few months :smiley:

**Pull request closed without merging**